### PR TITLE
Add linting, ES6 module and TypeScript support

### PR DESCRIPTION
**Pull request by HRKings**
Hey! First of all, sorry if this PR is a little intrusive.

This PR aims to modernize the package as a whole, what it does:

- Adds linting (provided by ESLint, using airbnb-base and custom rules)
- Converts to ES6 modules, this enables using `import/from`
- Adds pnpm support (a very fast and space efficient package manager, used by big packages like Vue)
- Updates the README to reflect the changes and apply some best practices.
- Add typescript support with JSDocs using the generator created by me and C-Garza based off [mudkipme](https://github.com/mudkipme/pokeapi-v2-typescript)

The underlying code logic has not changed and all tests are passing.

### CODE REVIEW

**Comment by HRKings** [comment]
Thanks!

Well, converting to a module requires that the person using also changes the type to module in the `package.json`, but all modern libraries that I looked up the source code uses the `"type": "module`, including node-fetch that is what the type generator uses

File: generator.tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+  "compilerOptions": {
+    "target": "esnext",
```
[the remark this answers is not in the document]
Comment (HRKings) [reply generator.tsconfig.json]: @C-Garza I moved the generators to ES2020. This makes the generators compatible with any LTS version of Node

File: generator/Utils.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import { fileURLToPath } from 'url';
+import path, { dirname } from 'path';
+
+export const __filename = fileURLToPath(import.meta.url);
+export const __dirname = dirname(__filename);
+
+// Paths that will be used
+export const typeFile = path.resolve(__dirname, '../../types/index.d.ts');
+export const mainFile = path.resolve(__dirname, '../../src/index.ts');
+export const apiMapFile = path.join(__dirname, 'apiMap.json');
```
Comment (Naramsim) [inline generator/Utils.ts:11]: I think this file is git ignored. When I run `npm run generate:main` it complains that it doesn't exist.

File: package.json [hunk 3]
```diff
@@ -1,15 +1,24 @@
...
     "node": ">=10"
   },
-  "main": "src/index.js",
+  "main": "dist/src/index.js",
+  "typings": "types/index.d.ts",
   "scripts": {
     "doctoc": "doctoc .",
-    "test": "mocha --exit"
+    "test": "tsc; mocha ./dist/test/",
+    "generate:main": "tsc -p ./generator/; node ./dist/generator/Main.js",
```
Comment (Naramsim) [inline package.json:14]: We can change the `;` with and `&&`

**Review by Naramsim — CHANGES_REQUESTED** [review]

File: generator/TypesGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+  quicktype, InputData, JSONSchemaInput, FetchingJSONSchemaStore,
+} from 'quicktype-core';
+
+import {
+  apiMapFile, schemaFolder, typesLabel, typeFile,
+} from './Utils.js';
+
+console.log(schemaFolder);
+if (!fs.existsSync(schemaFolder)) {
+  console.log('Schemas not found, please clone the "PokeAPI/api-data" repository on the root of this project first and change all $refs from "/schema/v2" to "api-data/data/schema/v2")');
```
Comment (Naramsim) [inline generator/TypesGenerator.ts:15]: We should then add an `npm` script that does the cloning and the replacing. How did you do the replacing? With `sed`?

File: generator/AddJSDocs.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    await loadDocumentation(namespace, docName);
+  }
+
+  // Save the file
+  await file.save();
+})();
+
+// Timestamp
+console.timeEnd(jsdocsLabel);
+console.log('JSDocs added!');
```
Comment (Naramsim) [inline generator/AddJSDocs.ts:104]: When running `npm run generate:jsdocs`, the comments are always appended and not updated, leading to every comment being duplicated. Is it possible to update them?

**Review by Naramsim — CHANGES_REQUESTED** [review]

File: generator/AddJSDocs.ts (hunk 5 above)
Comment (HRKings) [reply generator/AddJSDocs.ts]: Done ! I've also prevented empty JSDocs

File: generator/TypesGenerator.ts (hunk 4 above)
Comment (HRKings) [reply generator/TypesGenerator.ts:15]: VSCode, a global replace

File: package.json (hunk 3 above)
Comment (HRKings) [reply package.json]: Done

File: generator/Utils.ts (hunk 2 above)
Comment (HRKings) [reply generator/Utils.ts:11]: The API Map is generated when you run the `generate:types` to always be updated with the lastest schema available on the directory

**Review by Naramsim — CHANGES_REQUESTED** [review]
When running `npm run generate:main`, `PokeAPI` class and export are appended and thus duplicated.

**Comment by HRKings** [comment]
> When running `npm run generate:main`, `PokeAPI` class and export are appended and thus duplicated. In the `index.d.ts` file

The intention was to encourage regenerate the types file every time, so the main class would be always updated. But I've made it overwrite anyways now

**Comment by Naramsim** [comment]
Ok @HRKings, we are close to merging this PR :) It was hard and I still need to look into the code. But for now, I'm asking you to change the version to `4.0.0-beta.0` so we can push it on NPM.

**Comment by Naramsim** [comment]
Ah, another thing. Would you like to join PokeAPI library contributors so that you can support and maintain this package?

**Comment by HRKings** [comment]
> Ok @HRKings, we are close to merging this PR :) It was hard and I still need to look into the code. But for now, I'm asking you to change the version to `4.0.0-beta.0` so we can push it on NPM.

Changed it! It's been a long road, take your time to look into the code

**Comment by HRKings** [comment]
> Ah, another thing. Would you like to join PokeAPI library contributors so that you can support and maintain this package?

Sure! I would love to help and support the package

**Comment by Naramsim** [comment]
@HRKings , I'm terribly sorry for the delay.

I merged the PR, made some changes, you can check my commits if you want, and published an NPM `beta` version: https://www.npmjs.com/package/pokedex-promise-v2

Maybe you can test if including the module in a real typescript application works well. Or we could add a `.ts` test to this repo.

**Comment by HRKings** [comment]
@Naramsim, no worries about the delay!
I saw the changes, thanks for picking them up. I've been testing the package in a TS app from the beginning and now that I transitioned for the npm beta everything is working as expected.
Thanks for all the support!

**Comment by Naramsim** [comment]
Nice! Alright, so I added automated tests, saw that on Node12 Typescript transpiles well and the tests are all passing. In your opinion, can we set the `.engine` property in the `package.json` to `>=12`?

**Comment by HRKings** [comment]
I think we totally should set the minimum node version to 12, since it's the LTS and I can't really guarantee it will work on versions lower than that

**Pull request merged**